File: leet/0142_linked_list_cycle_2/0142_linked_list_cycle_2.ts
// LeetCode 142. Linked List Cycle II
// https://leetcode.com/problems/linked-list-cycle-ii/
// O(n) time, O(1) space
function detectCycle(head: ListNode | null): ListNode | null {
    let slow = head, fast = head;
    while (fast && fast.next) {
        fast = fast.next.next;
        slow = slow.next;
        if (slow == fast) {
            slow = head;
            while (slow != fast) {
                slow = slow.next;
                fast = fast.next;
            }
            return slow;
        }
    }
    return null;
};

File: leet/0142_linked_list_cycle_2/0142_linked_list_cycle_2.py
# LeetCode 142. Linked List Cycle II
# https://leetcode.com/problems/linked-list-cycle-ii/
# O(n) time, O(1) space
class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        slow, fast = head, head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                break
        if not fast or not fast.next:
            return None
        slow = head;
        while fast != slow:
            slow = slow.next
            fast = fast.next
        return slow

